// hd-utils/lib/paid_api.js
const util   = require('util')
const https = require("https")
const http = require("http")
const crypto = require('crypto')

let sendHttpPostRequest = function (path, content, callback) {
    let options = {  
        method: "POST",  
        host: "ggsdk.ichtx.com",  
        port: 81,  
        path: path,
        headers: {  
            "Content-Type": 'application/json',  
        }  
    };  
    //console.log(options)
    let reqq = http.request(options, function(ress){
        var resdata = ""
        ress.setEncoding('utf8');
        ress.on('data', function(data){
            resdata += data
        });
        ress.on('error', function(err){
            callback(err, null);
        });
        ress.on('end', function(err){
            callback(null, resdata);
        });
    });

    reqq.on('error', function(err){
        callback(err,null);
    });

    reqq.write(content);
    reqq.end();
}

let sendHttpsPostRequest = function (path, content, callback, host) {
    let ghost = host
    if(!ghost){
        ghost = "ggsdk.ichtx.com";
    }
    let options = {  
        method: "POST",  
        host: ghost,  
        port: 444,  
        path: path,
        rejectUnauthorized : false,
        headers: {  
            "Content-Type": 'application/json',  
        }  
    };  
    //console.log(options)
    let reqq = https.request(options, function(ress){
        var resdata = ""
        ress.setEncoding('utf8');
        ress.on('data', function(data){
            resdata += data
        });
        ress.on('error', function(err){
            callback(err, null);
        });
        ress.on('end', function(err){
            callback(null, resdata);
        });
    });

    reqq.on('error', function(err){
        callback(err,null);
    });

    reqq.write(content);
    reqq.end();
}
//https://ggsdk.ichtx.com:444/GenOrderId_offline
class PaidAPI {
    constructor(){

    }

    genPaidUrl(pid, headimg, nickname, plat_id, callback){
        //let url = `http://ggsdk.ichtx.com:81/daifupay?headimg=${headimg}&nickname=${nickname}&account=${pid}&channelid=${plat_id}`;
        //let url = `http://ggsdk.ichtx.com:81/daifupay?headimg=ABC&nickname=${nickname}&account=${pid}&channelid=${plat_id}`;
        // console.log(url)
        // http.get(url, function(req, ress) {
        //     var json = '';  
        //     req.on('data', function(data) {
        //         json += data;  
        //     });  
        //     req.on('end', function(){  
        //         console.log(json)
        //         var jsonOBJ = JSON.parse(json);
        //         if (jsonOBJ.ret == 200) {
        //             callback(null, jsonOBJ.url);
        //         }
        //         else{
        //             callback("data error", null)
        //         }
        //     });  
        // }).on('error', (e) => {
        //     callback(e.message, null)
        // });

        let content = {
            headimg : headimg,
            nickname : nickname,
            account : pid, 
            channelid : plat_id,
        }

        //let content = `headimg=${headimg}&nickname=${nickname}&account=${pid}&channelid=${plat_id}`;
        sendHttpPostRequest("/daifupay", JSON.stringify(content), function(err, ret){
            //console.log(err)
            if(err) return callback(err, null);
            //console.log(ret)
            var jsonOBJ = JSON.parse(ret);
            if (jsonOBJ.ret == 200) {
                callback(null, jsonOBJ.url);
            }
            else{
                callback("data error", null)
            }
        })
    }

    baiduPayNotify(params, callback){
        let url = `http://ggsdk.ichtx.com:81/baiduh5?params=${params}`
        //console.log(url)
        http.get(url, function(req, ress) {
            var json = '';  
            req.on('data', function(data) {
                json += data;  
            });  
            req.on('end', function(){  
               // console.log(json)
                callback(null, json)
            });  
        }).on('error', (e) => {
            callback(e.message, null)
        });
    }

    genOrder(price, plat_id, productid, userid, ext_params, callback){
        // let content = {
        //     appid : 20001, 
        //     channelid : plat_id,
        //     cporderid : "h5order"+userid+(new Date().getTime()),
        //     paychannel : "baiduh5",
        //     productid : productid,
        //     price : price,
        //     userid : userid,
        // }

        let content = {
            appid : 20001, 
            channelid : plat_id,
            cporderid : "h5order"+userid+(new Date().getTime()),
            paychannel : "wxpay",
            price : price,
            userid : userid,
            productid : productid,
        }
        for(let key in ext_params){
            content[key] = ext_params[key]
        }

        sendHttpsPostRequest("/GenOrderId_offline", JSON.stringify(content), function(err, ret){
            console.log(err, ret)
            if(err) return callback(err, null);
            return callback(null, ret)
        },"106.75.132.201")
    }
}

module.exports = PaidAPI;
