// hd-utils/lib/toy_api.js
const util   = require('util')
const https = require("https")

let sendHttpsPostRequest = function (host, path, content, callback) {
    let options = {  
        method: "POST",  
        host: host,  
        port: 443,  
        path: path,
        headers: {  
            "Content-Type": 'application/x-www-form-urlencoded;',  
        }  
    };  

    let reqq = https.request(options, function(ress){
    	var resdata = ""
        ress.setEncoding('utf8');
        ress.on('data', function(data){
            resdata += data
        });
        ress.on('error', function(err){
            callback(err, null);
        });
        ress.on('end', function(err){
            callback(null, resdata);
        });
    });

    reqq.on('error', function(err){
        callback(err,null);
    });

    reqq.write(content);
    reqq.end();
}

class ToyAPI {
	constructor(live_cfg){
		this.sdk_appid = live_cfg.sdk_appid
		this.admin_identifier = live_cfg.admin_identifier
		this.admin_sig = live_cfg.admin_sig
	}

	sendToyPost(host, path, params, callback){
		let real_path = util.format("%s?sdkappid=%s&identifier=%s&usersig=%s&random=%d&contenttype=json",
			path,this.sdk_appid, this.admin_identifier, this.admin_sig,Math.floor(Math.random() * 1000));

		sendHttpsPostRequest(host, real_path, params, callback)
	}

	joinToRoom(pid, rid, callback){
		let params = {
	        UserId : pid,
	        GroupId : rid
	    }
	    this.sendToyPost("console.tim.qq.com", "/v4/ilvb_doll_catch/room_addr_v2", 
	        JSON.stringify(params), function(err, ret){
	        callback(err, ret);
	    });
	}

    set_prob(rid, ProbNumerator, ProbDenominator, callback){
        let params = {
            GroupId : rid,
            ProbNumerator : ProbNumerator,
            ProbDenominator : ProbDenominator
        }
        
        this.sendToyPost("console.tim.qq.com", "/v4/ilvb_doll_catch/prob_set", JSON.stringify(params), function(err, ret){
            if(err) return callback(err, null)
            callback(null, JSON.parse(ret))
        });
    }

    
    queryRoomBatch(rids){
        let params = {
            GroupIds : rids
        }
        this.sendToyPost("console.tim.qq.com", "/v4/ilvb_doll_catch/room_batch", JSON.stringify(params), function(err, ret){
            if(err) return console.log(err);
            console.log(ret)
        });
    }
}

module.exports = ToyAPI;
